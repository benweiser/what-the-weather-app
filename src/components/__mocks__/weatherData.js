export const mockWeatherData = {
  name: 'Test',
  coord: {
    lat: 23,
    lon: 32
  },
  weather: [
    {
      id: 500,
      description: 'Test description'
    }
  ],
  main: {
    humidity: 40,
    temp: 45,
    temp_min: 23,
    temp_max: 79
  },
  sys: {
    sunrise: '5:00',
    sunset: '8:00'
  },
  wind: {
    speed: 43,
    deg: 43
  }
};
