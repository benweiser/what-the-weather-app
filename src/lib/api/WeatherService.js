import axios from 'axios';

const API_CONFIG = {
  basePath: 'https://api.openweathermap.org/data/',
  version: '2.5',
  apiKey: '3d75d3fd22342bb892e1650a52382b25',
  units: 'imperial'
};

/**
 * Gets the current weather
 * @param {string} query - the query param for making this api request
 */
const getCurrentWeather = query => {
  axios
    .get(
      `${API_CONFIG.basePath}${API_CONFIG.version}/weather?q=${query}&units=${
        API_CONFIG.units
      }&APPID=${API_CONFIG.apiKey}`
    )
    .then(response => console.log(response));
};

export default getCurrentWeather;

console.log(getCurrentWeather('London'));
