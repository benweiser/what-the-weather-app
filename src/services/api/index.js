import {
  getWeatherByCity,
  getWeatherByCoords,
  getWeatherByZipCode
} from './WeatherService';
import { getFlickrPhotosByCoords } from './FlickrService';

const API = {
  getWeatherByZipCode,
  getWeatherByCoords,
  getWeatherByCity,
  getFlickrPhotosByCoords
};

export default API;
