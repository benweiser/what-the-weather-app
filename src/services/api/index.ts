import {
  getCurrentWeather,
  getWeatherByCity,
  getWeatherByCoords,
  getWeatherByZipCode
} from './WeatherService';
import { getFlickrPhotosByCoords } from './FlickrService';

const API = {
  getCurrentWeather,
  getFlickrPhotosByCoords,
  getWeatherByCoords,
  getWeatherByCity,
  getWeatherByZipCode
};

export default API;
