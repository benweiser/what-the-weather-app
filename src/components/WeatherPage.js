import React from "react";
import styled, { css } from "react-emotion";
import API from "../services/api";
import WeatherSearch from "./WeatherSearch";
import WeatherStats from "./WeatherStats";
import Loader from "./Loader";
import Paper from "@material-ui/core/Paper";
import { getRandomFlickrPhoto } from "../services/api/FlickrService";

const StyledWeatherStats = css`
  padding: 32px;
  max-width: 700px;
  margin: 0 auto;
`;

const StyledWeatherSearchWrapper = styled("div")`
  background: url(${props => props.background});
  background-repeat: no-repeat;
  background-size: cover;
  padding: 32px;
  margin: 0 auto;
`;

const StyledWeatherSearch = css`
  margin-bottom: 32px;
`;

export const weatherMap = location => {
  const searchType = {
    city: value => API.getWeatherByCity(value),
    coords: value => API.getWeatherByCoords(value),
    zip: value => API.getWeatherByZipCode(value)
  };
  return searchType[location.searchMethod](location.value);
};

class WeatherPage extends React.Component {
  state = {
    data: {},
    fetching: false
  };

  onFetchWeather = async location => {
    this.setState({ fetching: true });

    const weatherData = await weatherMap(location);

    const {
      data: { coord }
    } = weatherData;
    const photoData = await API.getFlickrPhotosByCoords(
      coord.lat,
      coord.lon,
      weatherData.data.name
    );

    this.setState({
      currentPhoto: getRandomFlickrPhoto(photoData),
      data: weatherData.data,
      fetching: false,
      searchMethod: location.searchMethod,
      photos: photoData
    });
  };

  render() {
    const { data, currentPhoto, searchMethod } = this.state;

    if (!data) {
      return null;
    }

    return (
      <StyledWeatherSearchWrapper background={currentPhoto}>
        <Paper elevation={1} className={StyledWeatherStats}>
          <WeatherSearch
            className={StyledWeatherSearch}
            onFetchWeather={this.onFetchWeather}
            searchMethod={searchMethod}
          />
          {this.state.fetching ? (
            <Loader />
          ) : (
            data.main && <WeatherStats data={data} />
          )}
        </Paper>
      </StyledWeatherSearchWrapper>
    );
  }
}

export default WeatherPage;
