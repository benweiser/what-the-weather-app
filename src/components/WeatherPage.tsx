import React, { useState, memo } from 'react';
import 'styled-components/macro';

import API from '../services/api';
import { loadImage } from '../utils/image';
import WeatherSearch, { WeatherSearchState, SearchType } from './WeatherSearch';
import WeatherStats from './WeatherStats';

// import Loader from './Loader';
import Paper from '@material-ui/core/Paper';
import { getRandomFlickrPhoto } from '../services/api/FlickrService';
import { CurrentWeatherStats } from '../services/api/WeatherService';
import Skeleton from './Skeleton';

interface WeatherPageState {
  data?: CurrentWeatherStats;
  isError?: boolean;
  isLoading?: boolean;
  fetching?: boolean;
  currentPhoto: string;
  searchMethod: SearchType;
  photos?: ReadonlyArray<any>;
}

const searchType: { [key: string]: (value: string) => any } = {
  city: (value: string) => API.getWeatherByCity(value),
  // coords: (value: string) => API.getWeatherByCoords(value),
  zip: (value: string) => API.getWeatherByZipCode(value)
};

export const mapSearchTypeToRequest = (
  state: WeatherSearchState
): CurrentWeatherStats => {
  const { searchMethod, value } = state;

  return searchType[searchMethod](value);
};

const WeatherPage = () => {
  const [state, setState] = useState<WeatherPageState>({
    currentPhoto: '',
    data: undefined,
    isError: false,
    isLoading: false,
    searchMethod: 'city'
  });
  const handleFetchCurrentWeather = async (location: WeatherSearchState) => {
    setState({ ...state, isError: false, isLoading: true });
    const weatherData = await mapSearchTypeToRequest(location);

    if (!weatherData) {
      setState({
        ...state,
        data: undefined,
        currentPhoto: '',
        isError: true,
        isLoading: false
      });
      return null;
    }
    const { lat, lon, locationName } = weatherData;

    const photoData = await API.getFlickrPhotosByCoords(lat, lon, locationName);

    const photo = await loadImage(getRandomFlickrPhoto(photoData));

    setState({
      currentPhoto: photo ? photo.src : '',
      data: weatherData,
      isError: false,
      isLoading: false,
      searchMethod: location.searchMethod,
      photos: photoData
    });

    return null;
  };

  const { data, isError, isLoading, currentPhoto, searchMethod } = state;

  return (
    <div
      css={`
        background: url(${currentPhoto});
        background-repeat: no-repeat;
        background-attachment: fixed;
        background-size: cover;
        padding: 80px;
        margin: 0 auto;
      `}
    >
      <Paper
        elevation={1}
        css={`
          padding: 32px;
          max-width: 700px;
          margin: 0 auto;
        `}
      >
        <WeatherSearch
          onFetchWeather={handleFetchCurrentWeather}
          searchMethod={searchMethod}
        />
        <Skeleton isLoading={!!isLoading}>
          {data && <WeatherStats data={data} />}
        </Skeleton>
        {isError && <div data-testid="error-text">An error occured</div>}
      </Paper>
    </div>
  );
};

export default memo(WeatherPage);
