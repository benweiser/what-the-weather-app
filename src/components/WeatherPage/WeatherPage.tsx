import React from 'react';
// import styled, { css } from 'react-emotion';
import styled from 'styled-components/macro';
import API from '../../services/api';
import { loadImage } from '../../utils/image';
import WeatherSearch, {
  WeatherSearchState,
  SearchType
} from '../WeatherSearch';
import WeatherStats from '../WeatherStats';
import Loader from '../Loader';
import Paper from '@material-ui/core/Paper';
import { getRandomFlickrPhoto } from '../../services/api/FlickrService';
import { CurrentWeatherStats } from '../../services/api/WeatherService';

interface WeatherPageState {
  data?: CurrentWeatherStats;
  error?: boolean;
  fetching?: boolean;
  currentPhoto: string;
  searchMethod: SearchType;
  photos?: ReadonlyArray<any>;
}

interface WeatherPageProps {}

/*
const StyledWeatherStats = css`
  padding: 32px;
  max-width: 700px;
  margin: 0 auto;
`;
/*

const StyledWeatherSearchWrapper = styled('div')<{ background: string }>`
  background: url(${props => props.background});
  background-repeat: no-repeat;
  background-size: cover;
  padding: 32px;
  margin: 0 auto;
`;

/*
const StyledWeatherSearch = css`
  margin-bottom: 32px;
`;
*/

export const weatherMap = (
  location: WeatherSearchState
): CurrentWeatherStats => {
  const searchType: { [key: string]: (value: string) => any } = {
    city: (value: string) => API.getWeatherByCity(value),
    // coords: (value: string) => API.getWeatherByCoords(value),
    zip: (value: string) => API.getWeatherByZipCode(value)
  };
  return searchType[location.searchMethod](location.value);
};

class WeatherPage extends React.Component<WeatherPageProps, WeatherPageState> {
  state: WeatherPageState = {
    currentPhoto: '',
    data: undefined,
    error: false,
    fetching: false,
    searchMethod: 'city'
  };

  handleFetchCurrentWeather = async (location: WeatherSearchState) => {
    this.setState({ fetching: true });

    const weatherData = await weatherMap(location);

    if (!weatherData) {
      this.setState({ error: true, fetching: false });
      return null;
    }

    const { lat, lon, locationName } = weatherData;

    const photoData = await API.getFlickrPhotosByCoords(lat, lon, locationName);

    const photo = await loadImage(getRandomFlickrPhoto(photoData));

    this.setState({
      currentPhoto: photo.src,
      data: weatherData,
      error: false,
      fetching: false,
      searchMethod: location.searchMethod,
      photos: photoData
    });
  };

  render() {
    const { data, error, fetching, currentPhoto, searchMethod } = this.state;

    return (
      <div
        css={`
          background: url(${currentPhoto});
          border: 5px solid red;
        `}
      >
        <Paper elevation={1}>
          <WeatherSearch
            //  className={StyledWeatherSearch}
            onFetchWeather={this.handleFetchCurrentWeather}
            searchMethod={searchMethod}
          />
          {fetching ? <Loader /> : data && <WeatherStats data={data} />}
          {error && !fetching && (
            <div data-testid="error-text">An error occured</div>
          )}
        </Paper>
      </div>
    );
  }
}

export default WeatherPage;
