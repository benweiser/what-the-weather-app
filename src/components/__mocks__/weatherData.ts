import { CurrentWeatherStats } from './../../services/api/WeatherService';

export const mockRawWeatherData = {
  name: 'Test',
  coord: {
    lat: 23,
    lon: 32
  },
  weather: [
    {
      id: 500,
      icon: 'xyz',
      description: 'Test description',
      main: 'Test'
    }
  ],
  main: {
    humidity: 40,
    temp: 45,
    temp_min: 23,
    temp_max: 79
  },
  sys: {
    sunrise: 1571500519351,
    sunset: 1571500619351
  },
  wind: {
    speed: 43,
    deg: 43
  }
};

export const mockCurrentWeatherStats: CurrentWeatherStats = {
  currentConditions: [
    {
      description: 'Test description',
      icon: 'xyz',
      id: 500,
      main: 'Test'
    }
  ],
  lat: 23,
  locationName: 'Test',
  lon: 32,
  sunriseTime: 1571500519351,
  sunsetTime: 1571500619351,
  temp: 45,
  tempMax: 79,
  tempMin: 23,
  windSpeed: 43
};
