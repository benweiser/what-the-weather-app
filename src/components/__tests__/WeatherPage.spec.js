import { fireEvent, render, waitForElement } from 'react-testing-library';
import React from 'react';
import mockAxios from 'axios';

import WeatherPage from '../WeatherPage';
import { mockWeatherData } from '../__mocks__/weatherData';
import { API_CONFIG } from '../../services/api/WeatherService';

const setup = () => render(<WeatherPage />);

describe('Weather Page Component', () => {
  let consoleErrorSpy;
  beforeEach(() => {
    consoleErrorSpy = jest
      .spyOn(global.console, 'error')
      .mockImplementationOnce(() => {});
  });

  afterEach(() => {
    consoleErrorSpy.mockRestore();
    jest.resetAllMocks();
  });
  it('should render a weather page', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });

  it('should perform a weather search', async done => {
    const response = {
      data: {
        ...mockWeatherData
      }
    };
    mockAxios.get.mockImplementationOnce(() => Promise.resolve(response));
    const { getByTestId } = setup();
    fireEvent.click(getByTestId('search-button'));
    expect(mockAxios.get).toHaveBeenCalledTimes(1);
    await expect(mockAxios.get).toHaveBeenCalledWith('/weather?q=', API_CONFIG);
    await mockAxios.get.mockImplementationOnce(() =>
      Promise.resolve({
        data: {
          photos: {
            photo: {
              farm: 'test123',
              id: 123455,
              secret: 'e3sk3ksls9',
              server: 3432
            }
          }
        }
      })
    );
    expect(mockAxios.get).toHaveBeenCalledTimes(1);

    const weatherStatsNode = await waitForElement(() =>
      getByTestId('WeatherStats')
    );

    expect(weatherStatsNode).toHaveTextContent('Current Temperature:');
    done();
  });

  it('should handle errors on a failed search', async done => {
    mockAxios.get.mockImplementation(() => Promise.reject('search failed'));
    const { getByTestId } = setup();
    fireEvent.click(getByTestId('search-button'));
    expect(mockAxios.get).toHaveBeenCalledTimes(1);

    const errorContent = await waitForElement(() => getByTestId('error-text'));
    expect(consoleErrorSpy).toHaveBeenCalledTimes(1);
    expect(consoleErrorSpy).toHaveBeenCalledWith(
      'getAjax method failed',
      'search failed'
    );
    expect(errorContent).toHaveTextContent('An error occured');
    done();
  });
});
