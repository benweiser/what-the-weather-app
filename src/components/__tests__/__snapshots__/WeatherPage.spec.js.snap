// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Weather Page Component renders 1`] = `
<div
  className="css-1f77bxc"
>
  <div
    className="MuiPaper-root-1 MuiPaper-elevation1-4 MuiPaper-rounded-2 css-12128pa"
  >
    <div
      className="css-jtawwa"
    >
      <form
        onSubmit={[Function]}
      >
        <div
          aria-label="Select zip, coordinates, or location to search"
          className="MuiFormGroup-root-28 css-xbtlj5"
          role="radiogroup"
        >
          <label
            className="MuiFormControlLabel-root-30"
          >
            <span
              className="MuiButtonBase-root-49 MuiIconButton-root-43 MuiSwitchBase-root-39 MuiRadio-root-34 MuiRadio-colorSecondary-38 MuiSwitchBase-checked-40 MuiRadio-checked-35"
              onBlur={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              role={undefined}
              tabIndex={null}
            >
              <span
                className="MuiIconButton-label-48"
              >
                <svg
                  aria-hidden="true"
                  className="MuiSvgIcon-root-52"
                  color={undefined}
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 7c-2.76 0-5 2.24-5 5s2.24 5 5 5 5-2.24 5-5-2.24-5-5-5zm0-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                  />
                </svg>
                <input
                  autoFocus={undefined}
                  checked={true}
                  className="MuiSwitchBase-input-42"
                  disabled={undefined}
                  id={undefined}
                  name="searchby"
                  onChange={[Function]}
                  readOnly={undefined}
                  required={undefined}
                  tabIndex={undefined}
                  type="radio"
                  value="city"
                />
              </span>
              <span
                className="MuiTouchRipple-root-59"
              />
            </span>
            <span
              className="MuiTypography-root-66 MuiTypography-body1-75 MuiFormControlLabel-label-33"
            >
              By City
            </span>
          </label>
          <label
            className="MuiFormControlLabel-root-30"
          >
            <span
              className="MuiButtonBase-root-49 MuiIconButton-root-43 MuiSwitchBase-root-39 MuiRadio-root-34 MuiRadio-colorSecondary-38"
              onBlur={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              role={undefined}
              tabIndex={null}
            >
              <span
                className="MuiIconButton-label-48"
              >
                <svg
                  aria-hidden="true"
                  className="MuiSvgIcon-root-52"
                  color={undefined}
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                  />
                </svg>
                <input
                  autoFocus={undefined}
                  checked={false}
                  className="MuiSwitchBase-input-42"
                  disabled={undefined}
                  id={undefined}
                  name="searchby"
                  onChange={[Function]}
                  readOnly={undefined}
                  required={undefined}
                  tabIndex={undefined}
                  type="radio"
                  value="coords"
                />
              </span>
              <span
                className="MuiTouchRipple-root-59"
              />
            </span>
            <span
              className="MuiTypography-root-66 MuiTypography-body1-75 MuiFormControlLabel-label-33"
            >
              By Latitude/Longitude
            </span>
          </label>
          <label
            className="MuiFormControlLabel-root-30"
          >
            <span
              className="MuiButtonBase-root-49 MuiIconButton-root-43 MuiSwitchBase-root-39 MuiRadio-root-34 MuiRadio-colorSecondary-38"
              onBlur={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              role={undefined}
              tabIndex={null}
            >
              <span
                className="MuiIconButton-label-48"
              >
                <svg
                  aria-hidden="true"
                  className="MuiSvgIcon-root-52"
                  color={undefined}
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                  />
                </svg>
                <input
                  autoFocus={undefined}
                  checked={false}
                  className="MuiSwitchBase-input-42"
                  disabled={undefined}
                  id={undefined}
                  name="searchby"
                  onChange={[Function]}
                  readOnly={undefined}
                  required={undefined}
                  tabIndex={undefined}
                  type="radio"
                  value="zip"
                />
              </span>
              <span
                className="MuiTouchRipple-root-59"
              />
            </span>
            <span
              className="MuiTypography-root-66 MuiTypography-body1-75 MuiFormControlLabel-label-33"
            >
              By Zip Code
            </span>
          </label>
        </div>
        <div
          aria-describedby={undefined}
          aria-label="Enter city name"
          className="MuiFormControl-root-91 css-e8diyq"
        >
          <label
            className="MuiFormLabel-root-100 MuiInputLabel-root-95 MuiInputLabel-formControl-96 MuiInputLabel-animated-99"
            data-shrink={false}
            htmlFor={undefined}
          >
            Enter city name
          </label>
          <div
            className="MuiInput-root-107 MuiInput-formControl-108 MuiInput-underline-111"
          >
            <input
              aria-invalid={false}
              autoComplete={undefined}
              autoFocus={undefined}
              className="MuiInput-input-115"
              defaultValue={undefined}
              disabled={false}
              id={undefined}
              name={undefined}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={undefined}
              onKeyUp={undefined}
              placeholder="Enter city name"
              readOnly={undefined}
              required={false}
              rows={undefined}
              type="text"
              value=""
            />
          </div>
        </div>
        <button
          className="MuiButtonBase-root-49 MuiButton-root-120 MuiButton-contained-131 MuiButton-containedPrimary-132 MuiButton-raised-134 MuiButton-raisedPrimary-135"
          disabled={false}
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="submit"
        >
          <span
            className="MuiButton-label-121"
          >
            Get Weather
          </span>
          <span
            className="MuiTouchRipple-root-59"
          />
        </button>
      </form>
    </div>
  </div>
</div>
`;
